extensions/amp-analytics/0.1/cross-domain-iframe-message-queue.js

```diff
@@ -0,0 +1,164 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {dev} from '../../../src/log';
+import {timerFor} from '../../../src/services';
+import {
+  AMP_ANALYTICS_3P_MESSAGE_TYPE,
+} from '../../../src/3p-analytics-common';
+
+/** @private @const {string} */
+const TAG_ = 'amp-analytics.CrossDomainIframeMessageQueue';
+
+/** @private @const {number} */
+const MESSAGE_THROTTLE_TIME_ = 100;
+
+/** @private @const {number} */
+const MAX_QUEUE_SIZE_ = 100;
+
+/**
+ * @visibleForTesting
+ */
+export class CrossDomainIframeMessageQueue {
+  /**
+   * Constructor
+   * @param {!../../../3p/iframe-messaging-client.IframeMessagingClient}
+   *   iframeMessagingClient Facilitates
+   * cross-frame communication
+   * @param {!string} envelopeType The type of envelope that will be used to
+   * contain the queued messages (see 3p-analytics-common)
+   */
+  constructor(iframeMessagingClient, envelopeType) {
+    /** @type {Array<(
+    *            ../../../src/3p-analytics-common.ampAnalytics3pEvent|
+    *            ../../../src/3p-analytics-common.ampAnalytics3pNewCreative
+    *          )>} */
+    this.queue_ = [];
+
+    /** @type {?number} */
+    this.timerId_ = null;
+
+    /** @type {boolean} */
+    this.isReady_ = false;
+
+    /** @type {!../../../3p/iframe-messaging-client.IframeMessagingClient} */
+    this.iframeMessagingClient_ = iframeMessagingClient;
+
+    /** @type {string} */
+    this.envelopeType_ = envelopeType;
+  }
+
+  /**
+   * Destructor
+   */
+  destroy() {
+    CrossDomainIframeMessageQueue.timer_.cancel(this.timerId_);
+    this.timerId_ = null;
+  }
+
+  /**
+   * Indicate that a cross-domain frame is ready to receive messages, and
+   * send all messages that were previously queued for it.
+   * associated data
+   */
+  setIsReady() {
+    this.isReady_ = true;
+    this.sendQueuedMessages_();
+  }
+
+  /**
+   * Builds an event message to be passed to the cross-domain iframe(s)
+   * @param {!string} senderId
+   * @param {!string} message The data to send
+   * @return {../../../src/3p-analytics-common.ampAnalytics3pEvent}
+   */
+  static buildEventMessage(senderId, message) {
+    const messageObject = {
+      senderId,
+      type: AMP_ANALYTICS_3P_MESSAGE_TYPE.EVENT,
+    };
+    messageObject[AMP_ANALYTICS_3P_MESSAGE_TYPE.EVENT] = message;
+    const typedMessageObject =
+      /** @type {../../../src/3p-analytics-common.ampAnalytics3pEvent} */
+      (messageObject);
+    return typedMessageObject;
```

Comment:
Did it this way because if you do:
`
return
/** @type {../../../src/3p-analytics-common.ampAnalytics3pEvent} */
       (messageObject);
`
...then it interprets the "return" as a complete statement with no semicolon, and the subsequent two lines are unreachable code.